reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,258 @@
+
+- @jjyao
+- @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+- @jjyao
+
+## Design and Architecture
+
+### Prerequisites
+
+- The user have an active Spark cluster (apache/spark >= 3.3)
+
+- The user must install Ray packages on every node of the Spark cluster, e.g. via:
+
+  ```
+  pip install ray
+  ```
+
+  If the user's workload requires Ray add-ons, they must also be installed, e.g. via
+
+  ```
+  pip install "ray[debug,dashboard,tune,rllib,serve]"
+  ```
+
+### How to setup Ray cluster over spark cluster ?
+
+The architecture of a Spark cluster is as follows:
+![spark-cluster-overview](https://spark.apache.org/docs/latest/img/cluster-overview.png)
+
+Using Spark's barrier mode, we intend to represent each Ray worker node as a long-running Spark
+task, where the Ray worker node has the same set of resources as the Spark task. Multiple Spark
+tasks can run on a single Spark worker node, which means that multiple Ray worker nodes can run
+on a single Spark worker node. Finally, we intend to run the Ray head node on the Spark driver
+node with a fixed resource allocation.
+
+To clarify further, the following example demonstrates how a Ray cluster with 16 total worker CPU
+cores can be launched on a Spark cluster, assuming that each Ray node requires 4 CPU cores and
+10GB of memory:
+
+1. On the Spark driver node, launch the Ray head node as follows:
+
+```
+ray start --head --num-cpus=0
+```
+
+The Ray head node will operate solely as a  manager node, without worker node functionality.
+
+2. Create a Spark barrier mode job, which executes all tasks in the spark job concurrently. Each
+task is allocated a fixed number of CPUs and a fixed amount of memory. In this example, we will
+allocate 4 CPU cores to each Spark task and at least 10GB of memory (by ensuring that each Spark
+worker node has at least 10GB of memory reserved for every set of 4 CPU cores, as computed by
+``SPARK_WORKER_MEMORY`` / ``TASK_SLOTS_PER_WORKER`` under the assumption that the
+``spark.task.cpus`` configuration is set to ``4``).
+
+3. In each constituent Spark task, launch one Ray worker node in blocking mode and allocate to it
+the full set of resources available to the Spark task (4 CPUs, 10GB of memory). Keep the Spark task
+running until the Ray cluster destroyed. The command used to start each Ray worker node is as
+follows:
+
+```
+ray start —head --num-cpus=X --num-gpus=Y --memory=Z --address={head_ip}:{port}
+```
+
+4. After the Ray cluster is launched, the user's Ray application(s) can be submitted to the Ray
+cluster via the Ray head node address / port.
+
+5. To shut down the Ray cluster, cancel the Spark barrier mode job and terminate all Ray node
+services that are running on the Spark driver and Spark workers.
+
+Finally, by adjusting the resource allocation for Ray nodes and Spark tasks, this approach enables
+users to run multiple Ray clusters in isolation of each other on a single Spark cluster, if desired.
+
```

Comment:
Let's also describe fault tolerance as well: what will happen if a spark task (ray worker node) fails.

Reply:
I'll add a paragraph about the lack of fault tolerance in barrier execution mode and how entire tasks would need to retry based on the BarrierRDD implementation in Spark.

Reply:
yes barrier mode spark job lack of fault tolerance,
we can add a loop with retry logic inside BarrierRDD mapper function as the workaround.
Note the workaround does not support fault tolerance when the physical machine crashed.